test_cases/stdlib/check_dataclasses.py

```diff
@@ -55,17 +57,23 @@ def test_other_isdataclass_overloads(x: type, y: object) -> None:
     dc.replace(y)  # type: ignore
 
     if dc.is_dataclass(x):
+        assert_type(x, Type["DataclassInstance"])
         assert_type(dc.fields(x), Tuple[dc.Field[Any], ...])
-        # These should cause type checkers to emit errors
-        # due to the fact it's a dataclass class, not an instance
-        dc.asdict(x)  # type: ignore
-        dc.astuple(x)  # type: ignore
-        dc.replace(x)  # type: ignore
+
+        # Mypy correctly emits an error on these due to the fact
+        # that it's a dataclass class, not a dataclass instance.
+        # Pyright, however, handles ClassVar members in protocols differently.
+        # See https://github.com/microsoft/pyright/issues/4339
+        #
+        # dc.asdict(x)
+        # dc.astuple(x)
+        # dc.replace(x)
```

Comment:
We could possibly put these tests inside an `if MYPY` block, and add `"defineConstant": { "MYPY": false } }` to our pyrightconfig file for our test cases, to ensure that pyright doesn't check any code inside `if MYPY` blocks. I'm sort-of wary of doing that, however, as I'd really like to keep typechecker-specific test cases to a minimum in typeshed.